File: app/src/main/java/com/OdiousPanda/rainbow/Utilities/TextUtil.java
package com.OdiousPanda.rainbow.Utilities;

import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.net.Uri;
import android.text.Html;
import android.text.SpannableStringBuilder;
import android.text.TextPaint;
import android.text.style.ClickableSpan;
import android.text.style.URLSpan;
import android.view.View;

import androidx.annotation.NonNull;

import com.OdiousPanda.rainbow.API.APIConstants;
import com.OdiousPanda.rainbow.R;

import java.util.Random;

public class TextUtil {
    private static final String referralString = "?utm_source=rainbow&utm_medium=referral";

    public static SpannableStringBuilder getReferralHtml(Context context, String profileUrl, String profileName) {
        String html = context.getString(R.string.photo_by) + " <a href='";
        html += profileUrl + referralString;
        html += "'>";
        html += profileName;
        html += "</a> " + context.getString(R.string.on) + " <a href='";
        html += APIConstants.UNSPLASH_HOME_URL + referralString;
        html += "'> Unsplash</a>";
        CharSequence sequence = Html.fromHtml(html);
        SpannableStringBuilder strBuilder = new SpannableStringBuilder(sequence);
        URLSpan[] urls = strBuilder.getSpans(0, sequence.length(), URLSpan.class);
        for (URLSpan span : urls) {
            makeLinkClickable(context, strBuilder, span);
        }
        return strBuilder;
    }

    private static void makeLinkClickable(final Context context, SpannableStringBuilder strBuilder, final URLSpan span) {
        int start = strBuilder.getSpanStart(span);
        int end = strBuilder.getSpanEnd(span);
        int flags = strBuilder.getSpanFlags(span);
        ClickableSpan clickable = new ClickableSpan() {
            public void onClick(@NonNull View view) {
                Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(span.getURL()));
                context.startActivity(browserIntent);
            }

            @Override
            public void updateDrawState(@NonNull TextPaint ds) {
                super.updateDrawState(ds);
                ds.setColor(Color.BLUE);
                ds.setUnderlineText(true);
                //ds.setColor(Color.rgb(146,215,237));
            }
        };
        strBuilder.setSpan(clickable, start, end, flags);
        strBuilder.removeSpan(span);
    }

    public static String getTimeStringPretty(int h, int m) {
        String AP = " AM";
        String minute = m < 10 ? "0" + m : String.valueOf(m);
        if (h >= 12) {
            AP = " PM";
            if (h >= 13) {
                h -= 12;
            }
        }
        String hour = String.valueOf(h);
        return hour + ":" + minute + AP;
    }

    public static String locationStringForNearbySearch(String lat, String lon) {
        return lat + "," + lon;
    }

    static String getNotificationText(Context context, float temp, String summary, String precipitationType, float precipitationProb) {
        String notificationText = summary;
        String precipitationProbText = (int) (100 * precipitationProb) + "%";
        String additionalComment;
        String precipitationText = " There will be ";
        if (precipitationType != null) {
            precipitationText += precipitationProbText + " chance of " + precipitationType + ".";
            if (precipitationType.equals("rain")) {
                precipitationText += precipitationProb < 0.5 ? "" : " Bring an umbrella, or a raincoat.";
            }
        } else {
            precipitationText = "";
        }

        if (temp < 16) {
            String[] coldComments = context.getResources().getStringArray(R.array.cold_notification_comment);
            additionalComment = " " + coldComments[new Random().nextInt(coldComments.length)];
        } else if (temp > 30) {
            String[] hotComments = context.getResources().getStringArray(R.array.hot_notification_comment);
            additionalComment = " " + hotComments[new Random().nextInt(hotComments.length)];
        } else {
            String[] randomComments = context.getResources().getStringArray(R.array.random_notification_comment);
            additionalComment = " " + randomComments[new Random().nextInt(randomComments.length)];
        }

        notificationText += precipitationText + additionalComment;
        return notificationText;
    }

    public static String capitalizeSentence(String sentence) {
        sentence = sentence.toLowerCase();
        StringBuilder result = new StringBuilder();
        boolean capitalize = true;
        for (char c : sentence.toCharArray()) {
            if (capitalize) {
                result.append(Character.toUpperCase(c));
                if (!Character.isWhitespace(c) && c != '.') {
                    capitalize = false;
                }
            } else {
                result.append(c);
                if (c == '.') {
                    capitalize = true;
                }
            }
        }
        return result.toString();
    }
}

File: app/src/main/java/com/odiousPanda/rainbowKt/utilities/TextUtil.kt
package com.odiousPanda.rainbowKt.utilities

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.text.SpannableStringBuilder
import android.text.TextPaint
import android.text.style.ClickableSpan
import android.text.style.URLSpan
import android.view.View
import androidx.core.content.ContextCompat
import androidx.core.text.HtmlCompat
import com.odiousPanda.rainbowKt.R
import com.odiousPanda.rainbowKt.apis.APIConstants
import java.util.*

object TextUtil {
    private const val referralString = "?utm_source=rainbow&utm_medium=referral"

    fun getReferralHtml(
        context: Context,
        profileUrl: String,
        profileName: String?
    ): SpannableStringBuilder? {
        val html = "${context.getString(R.string.photo_by)} " +
                "<a href='$profileUrl$referralString'>$profileName</a> " +
                "${context.getString(R.string.on)} " +
                "<a href='${APIConstants.UNSPLASH_HOME_URL}$referralString}'> Unsplash</a>"

        val sequence: CharSequence = HtmlCompat.fromHtml(html, HtmlCompat.FROM_HTML_MODE_LEGACY)
        val strBuilder = SpannableStringBuilder(sequence)
        val urls =
            strBuilder.getSpans(0, sequence.length, URLSpan::class.java)
        for (span in urls) {
            makeLinkClickable(context, strBuilder, span)
        }
        return strBuilder
    }

    private fun makeLinkClickable(
        context: Context,
        strBuilder: SpannableStringBuilder,
        span: URLSpan
    ) {
        val start = strBuilder.getSpanStart(span)
        val end = strBuilder.getSpanEnd(span)
        val flags = strBuilder.getSpanFlags(span)
        val clickable: ClickableSpan = object : ClickableSpan() {
            override fun onClick(view: View) {
                val browserIntent =
                    Intent(Intent.ACTION_VIEW, Uri.parse(span.url))
                context.startActivity(browserIntent)
            }

            override fun updateDrawState(ds: TextPaint) {
                super.updateDrawState(ds)
                ds.color = ContextCompat.getColor(context, R.color.hyperBlue)
                ds.isUnderlineText = true
            }
        }
        strBuilder.setSpan(clickable, start, end, flags)
        strBuilder.removeSpan(span)
    }

    fun getTimeStringPretty(h: Int, m: Int): String? {
        var hTemp = h
        var AP = " AM"
        val minute = if (m < 10) "0$m" else m.toString()
        if (hTemp >= 12) {
            AP = " PM"
            if (hTemp >= 13) {
                hTemp -= 12
            }
        }
        val hour = hTemp.toString()
        return "$hour:$minute$AP"
    }

    fun locationStringForNearbySearch(lat: String, lon: String): String {
        return "$lat,$lon"
    }

    fun getNotificationText(
        context: Context,
        temp: Float,
        summary: String,
        precipitationType: String,
        precipitationProb: Float
    ): String {
        var notificationText = summary
        val precipitationProbText = "${(100 * precipitationProb).toInt()}%"
        val additionalComment: String
        var precipitationText = " ${context.getString(R.string.there_will_be)} "
        precipitationText += "$precipitationProbText ${context.getString(R.string.chance_of)} $precipitationType."
        if (precipitationType == "rain") {
            precipitationText += if (precipitationProb < 0.5) "" else " ${context.getString(R.string.bring_umbrella)}"
        }
        val tempC = UnitConverter.toCelsius(temp)
        additionalComment = when {
            tempC < Constants.APT_TEMP_COLD -> {
                val coldComments =
                    context.resources.getStringArray(R.array.cold_notification_comment)

                " ${coldComments[Random().nextInt(coldComments.size)]}"
            }
            tempC > Constants.APT_TEMP_HOT -> {
                val hotComments =
                    context.resources.getStringArray(R.array.hot_notification_comment)

                " ${hotComments[Random().nextInt(hotComments.size)]}"
            }
            else -> {
                val randomComments =
                    context.resources.getStringArray(R.array.random_notification_comment)

                " ${randomComments[Random().nextInt(randomComments.size)]}"
            }
        }
        notificationText += "$precipitationText$additionalComment"
        return notificationText
    }

    fun capitalizeSentence(sentence: String): String {
        val sentenceTmp = sentence.toLowerCase(Locale.ROOT)
        val result = StringBuilder()
        var capitalize = true
        for (c in sentenceTmp.toCharArray()) {
            if (capitalize) {
                result.append(Character.toUpperCase(c))
                if (!Character.isWhitespace(c) && c != '.') {
                    capitalize = false
                }
            } else {
                result.append(c)
                if (c == '.') {
                    capitalize = true
                }
            }
        }
        return result.toString()
    }
}